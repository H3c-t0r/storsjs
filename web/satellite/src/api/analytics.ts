// Copyright (C) 2021 Storj Labs, Inc.
// See LICENSE for copying information.

import { HttpClient } from '@/utils/httpClient';

/**
 * AnalyticsHttpApi is a console Analytics API.
 * Exposes all analytics-related functionality
 */
export class AnalyticsHttpApi {
    private readonly http: HttpClient = new HttpClient();
    private readonly ROOT_PATH: string = '/api/v0/analytics';

    /**
     * Used to get authentication token.
     *
     * @param eventName - name of the event
     * @throws Error
     */
    public async eventTriggered(eventName: string): Promise<void> {
        const path = `${this.ROOT_PATH}/event`;
        const body = {
            eventName: eventName,
        };
        const response = await this.http.post(path, JSON.stringify(body));
        if (response.ok) {
            return;
        }

        throw new Error('Can not track event');
    }
}