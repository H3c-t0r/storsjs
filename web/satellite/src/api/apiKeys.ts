// Copyright (C) 2018 Storj Labs, Inc.
// See LICENSE for copying information.

import apollo from '@/utils/apolloManager';
import gql from 'graphql-tag';

export async function fetchAPIKeys(projectID: string) {
    let result: RequestResponse<any[]> = {
        errorMessage: '',
        isSuccess: false,
        data: []
    };

    try {
        let response: any = await apollo.query({
            query: gql(
                `query {
                    project(
                        id: "${projectID}",
                    ) {
                        apiKeys {
                            id,
                            name,
                            createdAt
                        }
                    }
                }`
            ),
            fetchPolicy: 'no-cache'
        });

        if (response.errors) {
            result.errorMessage = response.errors[0].message;
        } else {
            result.isSuccess = true;
            result.data = response.data.project.apiKeys;
        }
    } catch (e) {
        result.errorMessage = e.message;
    }

    return result;
}

export async function createAPIKey(projectID: string, name: string) {
    let result: RequestResponse<any> = {
        errorMessage: '',
        isSuccess: false,
        data: null
    };

    try {
        let response: any = await apollo.mutate({
            mutation: gql(
                `mutation {
                    createAPIKey(
                        projectID: "${projectID}",
                        name: "${name}"
                    ) {
                        key,
                        keyInfo {
                            id,
                            name,
                            createdAt
                        }
                    }
                }`
            ),
            fetchPolicy: 'no-cache'
        });

        if (response.errors) {
            result.errorMessage = response.errors[0].message;
        } else {
            console.log('response', response)
            result.isSuccess = true;
            result.data = {
                key: response.data.createAPIKey.key,
                keyInfo: response.data.createAPIKey.keyInfo
            }
        }
    } catch (e) {
        result.errorMessage = e.message;
    }

    return result;
}

export async function deleteAPIKey(id: string) {
    let result: RequestResponse<any> = {
        errorMessage: '',
        isSuccess: false,
        data: null
    };

    try {
        let response: any = await apollo.mutate({
            mutation: gql(
                `mutation {
                    deleteAPIKey(id: "${id}") {
                        id
                    }
                }`
            ),
            fetchPolicy: 'no-cache'
        });

        if (response.errors) {
            result.errorMessage = response.errors[0].message;
        } else {
            result.isSuccess = true;
            result.data = response.data.deleteAPIKey;
        }
    } catch (e) {
        result.errorMessage = e.message;
    }

    return result;
}