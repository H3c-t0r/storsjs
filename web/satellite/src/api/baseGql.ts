// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

import gql from 'graphql-tag';
import apolloManager from '@/utils/apollo';

/**
 * BaseGql is a graphql utility which allows to perform queries and mutations
 */
export class BaseGql {
    /**
     * performs qraphql query
     *
     * @param query - qraphql query
     * @param variables - variables to bind in query. null by default.
     * @throws Error
     */
    protected async query(query: string, variables: any = null): Promise<any> {
        let response: any = await apolloManager.query(
            {
                query: gql(query),
                variables,
                fetchPolicy: 'no-cache',
                errorPolicy: 'all',
            }
        );

        if (response.errors) {
            throw new Error(this.combineErrors(response.errors));
        }

        return response;
    }

    /**
     * performs qraphql mutation
     *
     * @param query - qraphql query
     * @param variables - variables to bind in query. null by default.
     * @throws Error
     */
    protected async mutate(query: string, variables: any = null): Promise<any> {
        let response: any = await apolloManager.mutate(
            {
                mutation: gql(query),
                variables,
                fetchPolicy: 'no-cache',
                errorPolicy: 'all',
            }
        );

        if (response.errors) {
            throw new Error(this.combineErrors(response.errors));
        }

        return response;
    }

    private combineErrors(gqlError: any): string {
        return gqlError.map(err => err.message).join('\n');
    }
}
