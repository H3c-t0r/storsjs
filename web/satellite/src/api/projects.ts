// Copyright (C) 2018 Storj Labs, Inc.
// See LICENSE for copying information.

import apollo from '@/utils/apolloManager';
import gql from 'graphql-tag';

// Performs graqhQL request.
// Throws an exception if error occurs
export async function createProject(project: Project): Promise<any> {
	let response = await apollo.mutate(
		{
			mutation: gql(`
                mutation {
                    createProject(
                        input: {
                            name: "${project.name}",
                            description: "${project.description}",
                            companyName: "${project.companyName}",
                            isTermsAccepted: ${project.isTermsAccepted},
                        }
                    ) {id}
                }`
			),
			fetchPolicy: 'no-cache',
		}
	);

	if (!response) {
		// TODO: replace with popup in future
		console.error('cannot create project');

		return null;
	}

	return response;
}

// Performs graqhQL request for fetching all projects of current user.
export async function fetchProjects(): Promise<any> {
	let response = await apollo.query(
		{
			query: gql(`
                query {
                    myProjects{
                        name
                        id
                        description
                        createdAt            
                        ownerName           
                        companyName           
                        isTermsAccepted
                    }
                }`
			),
			fetchPolicy: 'no-cache',
		}
	);

	if (!response) {
		// TODO: replace with popup in future
		console.error('cannot fetch projects');

		return null;
	}

	return response;
}

// Performs graqhQL request for updating selected project description
export async function updateProject(projectID: string, description: string): Promise<any> {
	let response = await apollo.mutate(
		{
			mutation: gql(`
                mutation {
                    updateProjectDescription(
                        id: "${projectID}",
                        description: "${description}"
                    )
                }`
			),
			fetchPolicy: 'no-cache',
		}
	);

	if (!response) {
		// TODO: replace with popup in future
		console.error('cannot update project');

		return null;
	}

	return response;
}

// Performs graqhQL request for deleting selected project
export async function deleteProject(projectID: string): Promise<any> {
	let response = await apollo.mutate(
		{
			mutation: gql(`
                mutation {
                    deleteProject(
                        id: "${projectID}"
                    )
                }`
			),
			fetchPolicy: 'no-cache',
		}
	);

	if (!response) {
		// TODO: replace with popup in future
		console.error('cannot delete project');

		return null;
	}

	return response;
}
