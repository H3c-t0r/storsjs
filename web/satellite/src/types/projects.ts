// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

/**
 * Exposes all project-related functionality
 */
export interface ProjectsApi {
    /**
     * Creates project
     *
     * @param createProjectModel - contains project information
     * @throws Error
     */
    create(createProjectModel: CreateProjectModel): Promise<Project>;
    /**
     * Fetch projects
     *
     * @returns Project[]
     * @throws Error
     */
    get(): Promise<Project[]>;
    /**
     * Update project
     *
     * @param projectId - project ID
     * @param description - project description
     * @returns Project[]
     * @throws Error
     */
    update(projectId: string, description: string): Promise<void>;
    /**
     * Delete project
     *
     * @param projectId - project ID
     * @throws Error
     */
    delete(projectId: string): Promise<void>;
}

// Project is a type, used for creating new project in backend
export class Project {
    public id: string;

    public name: string;
    public description: string;
    public createdAt: string;
    public ownerId: string;

    public isSelected: boolean;

    public constructor(id: string = '', name: string = '', description: string = '', createdAt: string = '', ownerId: string = '', isSelected: boolean = false) {
        this.id = id;
        this.name = name;
        this.description = description;
        this.createdAt = createdAt;
        this.isSelected = isSelected;
        this.ownerId = ownerId;
    }
}

// UpdateProjectModel is a type, used for updating project description
export class UpdateProjectModel {
    public id: string;
    public description: string;

    public constructor(id: string, description: string) {
        this.id = id;
        this.description = description;
    }
}

// CreateProjectModel is a type, used for creating project
export class CreateProjectModel {
    public name: string;
    public description: string;
}
