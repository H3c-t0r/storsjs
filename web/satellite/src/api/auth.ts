// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

import { BaseGql } from '@/api/baseGql';
import { User } from '@/types/users';

/**
 * AuthApiGql is a graphql implementation of Auth API.
 * Exposes all auth-related functionality
 */
export class AuthApi extends BaseGql {
    /**
     * Used to resend an registration confirmation email
     *
     * @param userId - id of newly created user
     * @throws Error
     */
    public async resendEmail(userId: string): Promise<void> {
        const query =
            `query ($userId: String!){
                resendAccountActivationEmail(id: $userId)
            }`;

        const variables = {
            userId,
        };

        await this.query(query, variables);
    }

    /**
     * Used to get authentication token
     *
     * @param email - email of the user
     * @param password - password of the user
     * @throws Error
     */
    public async token(email: string, password: string): Promise<string> {
        const query =
            ` query ($email: String!, $password: String!) {
                token(email: $email, password: $password) {
                    token
                }
            }`;

        const variables = {
            email,
            password,
        };

        const response = await this.query(query, variables);

        return response.data.token.token;
    }

    /**
     * Used to restore password
     *
     * @param email - email of the user
     * @throws Error
     */
    public async forgotPassword(email: string): Promise<void> {
        const query =
            `query($email: String!) {
                forgotPassword(email: $email)
            }`;

        const variables = {
            email,
        };

        await this.query(query, variables);
    }

    /**
     * Used to change password
     *
     * @param password - old password of the user
     * @param newPassword - new password of the user
     * @throws Error
     */
    public async changePassword(password: string, newPassword: string): Promise<void> {
        const query =
            `mutation($password: String!, $newPassword: String!) {
                changePassword (
                    password: $password,
                    newPassword: $newPassword
                ) {
                   email
                }
            }`;

        const variables = {
            password,
            newPassword,
        };

        await this.mutate(query, variables);
    }

    /**
     * Used to delete account
     *
     * @param password - password of the user
     * @throws Error
     */
    public async delete(password: string): Promise<void> {
        const query =
            `mutation ($password: String!){
                deleteAccount(password: $password) {
                    email
                }
            }`;

        const variables = {
            password,
        };

        await this.mutate(query, variables);
    }

    // TODO: remove secret after Vanguard release
    /**
     * Used to create account
     *
     * @param user - stores user information
     * @param secret - registration token used in Vanguard release
     * @param refUserId - referral id to participate in bonus program
     * @returns id of created user
     * @throws Error
     */
    public async create(user: User, password: string, secret: string, referrerUserId: string = ''): Promise<string> {
        const query =
            `mutation($email: String!, $password: String!, $fullName: String!, $shortName: String!,
                     $partnerID: String!, $referrerUserId: String!, $secret: String!) {
                createUser(
                    input: {
                        email: $email,
                        password: $password,
                        fullName: $fullName,
                        shortName: $shortName,
                        partnerId: $partnerID
                    },
                    referrerUserId: $referrerUserId,
                    secret: $secret,
                ) {email, id}
            }`;

        const variables = {
            email: user.email,
            fullName: user.fullName,
            shortName: user.shortName,
            partnerID: user.partnerId ? user.partnerId : '',
            referrerUserId: referrerUserId ? referrerUserId : '',
            password,
            secret,
        };

        const response = await this.mutate(query, variables);

        return response.data.createUser.id;
    }
}
