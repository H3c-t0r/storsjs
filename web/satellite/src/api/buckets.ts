// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

import { BaseGql } from '@/api/baseGql';
import { BucketCursor, BucketPage, BucketsApi } from '@/types/buckets';

/**
 * BucketsApiGql is a graphql implementation of Buckets API.
 * Exposes all bucket-related functionality
 */
export class BucketsApiGql extends BaseGql implements BucketsApi {
    /**
     * Fetch buckets
     *
     * @returns BucketPage
     * @throws Error
     */
    public async get(projectId: string, before: Date, cursor: BucketCursor): Promise<BucketPage> {
        const query =
            `query($projectId: String!, $before: DateTime!, $limit: Int!, $search: String!, $page: Int!) {
                project(id: $projectId) {
                    bucketUsages(before: $before, cursor: {
                        limit: $limit, search: $search, page: $page
                    }) {
                        bucketUsages {
                            bucketName,
                            storage,
                            egress,
                            objectCount,
                            since,
                            before
                        },
                        search,
                        limit,
                        offset,
                        pageCount,
                        currentPage,
                        totalCount
                    }
                }
            }`;

        const variables = {
            projectId,
            before: before.toISOString(),
            limit: cursor.limit,
            search: cursor.search,
            page: cursor.page,
        };

        const response = await this.query(query, variables);

        return this.fromJson(response.data.project.bucketUsages);
    }

    private fromJson(bucketPage): BucketPage {
        return new BucketPage(bucketPage.bucketUsages, bucketPage.search, bucketPage.limit, bucketPage.offset, bucketPage.pageCount, bucketPage.currentPage, bucketPage.totalCount);
    }
}
