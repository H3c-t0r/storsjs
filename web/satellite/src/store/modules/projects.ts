// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

import { StoreModule } from '@/store';
import { CreateProjectModel, Project, ProjectsApi, UpdateProjectModel } from '@/types/projects';

export const PROJECTS_ACTIONS = {
    FETCH: 'fetchProjects',
    CREATE: 'createProject',
    SELECT: 'selectProject',
    UPDATE: 'updateProject',
    DELETE: 'deleteProject',
    CLEAR: 'clearProjects',
};

export const PROJECTS_MUTATIONS = {
    ADD: 'CREATE_PROJECT',
    REMOVE: 'DELETE_PROJECT',
    UPDATE_PROJECT: 'UPDATE_PROJECT',
    SET_PROJECTS: 'SET_PROJECTS',
    SELECT_PROJECT: 'SELECT_PROJECT',
    CLEAR_PROJECTS: 'CLEAR_PROJECTS',
};

const defaultSelectedProject = new Project('', '', '', '', '', true);

class ProjectsState {
    public projects: Project[] = [];
    public selectedProject: Project = defaultSelectedProject;
}

const {
    FETCH,
    CREATE,
    SELECT,
    UPDATE,
    DELETE,
    CLEAR,
} = PROJECTS_ACTIONS;

const {
    ADD,
    REMOVE,
    UPDATE_PROJECT,
    SET_PROJECTS,
    SELECT_PROJECT,
    CLEAR_PROJECTS,
} = PROJECTS_MUTATIONS;

export function makeProjectsModule(api: ProjectsApi): StoreModule<ProjectsState> {
    return {
        state: new ProjectsState(),
        mutations: {
            [ADD](state: any, createdProject: Project): void {
                state.projects.push(createdProject);
            },
            [SET_PROJECTS](state: any, projects: Project[]): void {
                state.projects = projects;

                if (!state.selectedProject.id) {
                    return;
                }

                const projectsCount = state.projects.length;

                for (let i = 0; i < projectsCount; i++) {
                    const project = state.projects[i];

                    if (project.id !== state.selectedProject.id) {
                        continue;
                    }

                    state.selectedProject = project;

                    return;
                }

                state.selectedProject = defaultSelectedProject;
            },
            [SELECT_PROJECT](state: any, projectID: string): void {
                const selected = state.projects.find((project: any) => project.id === projectID);

                if (!selected) {
                    return;
                }

                state.selectedProject = selected;
            },
            [UPDATE_PROJECT](state: any, updateProjectModel: UpdateProjectModel): void {
                const selected = state.projects.find((project: any) => project.id === updateProjectModel.id);
                if (!selected) {
                    return;
                }

                selected.description = updateProjectModel.description;
            },
            [REMOVE](state: any, projectID: string): void {
                state.projects = state.projects.filter(project => project.id !== projectID);

                if (state.selectedProject.id === projectID) {
                    state.selectedProject = new Project();
                }
            },
            [CLEAR_PROJECTS](state: ProjectsState): void {
                state.projects = [];
                state.selectedProject = defaultSelectedProject;
            },
        },
        actions: {
            [FETCH]: async function ({commit}: any): Promise<Project[]> {
                const projects = await api.get();

                commit(SET_PROJECTS, projects);

                return projects;
            },
            [CREATE]: async function ({commit}: any, createProjectModel: CreateProjectModel): Promise<Project> {
                const project = await api.create(createProjectModel);

                commit(ADD, project);

                return project;
            },
            [SELECT]: function ({commit}: any, projectID: string): void {
                commit(SELECT_PROJECT, projectID);
            },
            [UPDATE]: async function ({commit}: any, updateProjectModel: UpdateProjectModel): Promise<void> {
                await api.update(updateProjectModel.id, updateProjectModel.description);

                commit(UPDATE_PROJECT, updateProjectModel);
            },
            [DELETE]: async function ({commit}: any, projectID: string): Promise<void> {
                await api.delete(projectID);

                commit(REMOVE, projectID);
            },
            [CLEAR]: function({commit}: any): void {
                commit(CLEAR_PROJECTS);
            },
        },
        getters: {
            projects: (state: any): Project[] => {
                return state.projects.map((project: any) => {
                    if (project.id === state.selectedProject.id) {
                        project.isSelected = true;
                    }

                    return project;
                });
            },
            selectedProject: (state: any): Project => state.selectedProject,
        },
    };
}
