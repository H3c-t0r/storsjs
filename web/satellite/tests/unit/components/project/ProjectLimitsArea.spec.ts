// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

import Vuex from 'vuex';

import ProjectLimitsArea from '@/components/project/ProjectLimitsArea.vue';

import { makeProjectsModule } from '@/store/modules/projects';
import { createLocalVue, shallowMount } from '@vue/test-utils';

import { ProjectsApiMock } from '../../mock/api/projects';

const localVue = createLocalVue();
localVue.use(Vuex);

const projectsApi = new ProjectsApiMock();
const projectsModule = makeProjectsModule(projectsApi);
const store = new Vuex.Store({ modules: { projectsModule }});

describe('ProjectLimitsArea', () => {
    it('snapshot not changed', () => {
        const wrapper = shallowMount(ProjectLimitsArea, {
            store,
            localVue,
        });

        expect(wrapper).toMatchSnapshot();
    });
});
