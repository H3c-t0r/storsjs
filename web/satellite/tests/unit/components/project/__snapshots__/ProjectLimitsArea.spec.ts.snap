// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectLimitsArea snapshot not changed 1`] = `
<div class="limits-container">
  <div class="limits-container__item">
    <p class="limits-container__item__title">Egress limits used</p>
    <div class="limits-container__item__values-container">
      <p class="limits-container__item__values-container__remaining">0</p>
      <p class="limits-container__item__values-container__total">/ 0</p>
    </div>
  </div>
  <div class="limits-container__item">
    <p class="limits-container__item__title">Storage limits used</p>
    <div class="limits-container__item__values-container">
      <p class="limits-container__item__values-container__remaining">0</p>
      <p class="limits-container__item__values-container__total">/ 0</p>
    </div>
  </div>
</div>
`;
